### PR TITLE
src: make OnWorkComplete and OnExecute override-able

### PR DESCRIPTION
**Pull request by legendecas**
No breaking changes on existing code were expected. All existing tests shall pass without any touch.
Changes on declaration:
- Added `Napi::AsyncWorker::OnWorkComplete`.
- Added `Napi::AsyncWorker::OnExecute`.

Fixes: #582 

- [x] npm test passed.
- [x] documentation has changed.

### CODE REVIEW

**Comment by legendecas** [comment]
`Napi::AsyncProgressWorkerBase` should be able to accept an option to set the queue size of the tsfn.

**Comment by legendecas** [comment]
There might be issues on sharing the AsyncProgressWorker base class between AsyncProgressWorker and AsyncProgressQueueWorker if the ThreadSafeFunction queue size is different: a possible redundant wrapper around ThreadSafeFunction Call might be introduced to AsyncProgressWorker.

So I'd like to remove the AsyncProgressWorker base class from the PR and leave the PR for the AsyncWorker's methods.

**Comment by mhdawson** [comment]
This one is ready to be reviewed @nodejs/n-api please review/comment.

File: doc/async_worker.md [hunk 1]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
 virtual void Napi::AsyncWorker::OnError(const Napi::Error& e);
 ```
 
+### OnWorkComplete
+
+This method is invoked after the work has completed on JavaScript thread.
+The default implementation of this method checks the status of the work and
+try to dispatch the result to `Napi::AsyncWorker::OnOk` or `Napi::AsyncWorker::Error`
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
tries to dispatch the result to `Napi::AsyncWorker::OnOk` or `Napi::AsyncWorker::Error`
```

File: doc/async_worker.md [hunk 2]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+if the work has committed an error. If the work was cancelled, neither of
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
if the work has committed an error. If the work was cancelled, neither
```

File: doc/async_worker.md [hunk 3]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+`Napi::AsyncWorker::OnOk` nor `Napi::AsyncWorker::Error` will be invoked.
+After the result dispatched, the default implementation will call into
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
After the result is dispatched, the default implementation will call into
```

File: doc/async_worker.md [hunk 4]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+`Napi::AsyncWorker::Destroy` if `SuppressDestruct()` was not called.
+
+```cpp
+virtual void OnWorkComplete(Napi::Env env, napi_status status);
+```
+
+### OnExecute
+
+This method is invoked immediately on the work thread on scheduled.
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
This method is invoked immediately on the work thread when scheduled.
```

File: doc/async_worker.md [hunk 5]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+The default implementation of this method just call the `Napi::AsyncWorker::Execute`
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
The default implementation of this method just calls the `Napi::AsyncWorker::Execute`
```

File: doc/async_worker.md [hunk 6]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+and handles exceptions if cpp exceptions was enabled.
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
and handles exceptions if cpp exceptions were enabled.
```

File: napi.h [hunk 7]
```diff
@@ -1943,6 +1943,12 @@ namespace Napi {
     napi_async_context _context;
   };
 
+
+  inline void OnAsyncWorkExecute(napi_env env, void* asyncworker);
+  inline void OnAsyncWorkComplete(napi_env env,
+                                  napi_status status,
+                                  void* asyncworker);
+
```
Comment (gabrielschulhof) [inline napi.h]: These should be part of the `AsyncWorker` class, in the private section.

File: doc/async_worker.md [hunk 8]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
 virtual void Napi::AsyncWorker::OnError(const Napi::Error& e);
 ```
 
+### OnWorkComplete
+
+This method is invoked after the work has completed on JavaScript thread.
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
This method is invoked after the work has completed on the JavaScript thread.
```

File: doc/async_worker.md [hunk 9]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+`Napi::AsyncWorker::Destroy` if `SuppressDestruct()` was not called.
+
+```cpp
+virtual void OnWorkComplete(Napi::Env env, napi_status status);
+```
+
+### OnExecute
+
+This method is invoked immediately on the work thread when scheduled.
+The default implementation of this method just calls the `Napi::AsyncWorker::Execute`
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
The default implementation of this method just calls the `Napi::AsyncWorker::Execute` method
```

File: doc/async_worker.md [hunk 10]
```diff
@@ -136,6 +136,30 @@ class was created, passing in the error as the first parameter.
...
+```cpp
+virtual void OnWorkComplete(Napi::Env env, napi_status status);
+```
+
+### OnExecute
+
+This method is invoked immediately on the work thread when scheduled.
+The default implementation of this method just calls the `Napi::AsyncWorker::Execute` method
+and handles exceptions if cpp exceptions were enabled.
+
```
Comment (gabrielschulhof) [inline doc/async_worker.md]: Maybe add a paragraph here warning sternly against using the `Napi::Env` inside `OnExecute`.

File: napi-inl.h [hunk 11]
```diff
@@ -4199,40 +4199,51 @@ inline void AsyncWorker::SetError(const std::string& error) {
 inline std::vector<napi_value> AsyncWorker::GetResult(Napi::Env /*env*/) {
   return {};
 }
+// The OnAsyncWorkExecute method receives an napi_env argument. However, do NOT
+// use it within this method, as it does not run on the main thread and must
+// not run any method that would cause JavaScript to run. In practice, this
+// means that almost any use of napi_env will be incorrect.
```
Comment (gabrielschulhof) [inline napi-inl.h]: This blurb should be copied into the documentation above.

File: napi-inl.h [hunk 12]
```diff
@@ -4199,40 +4199,51 @@ inline void AsyncWorker::SetError(const std::string& error) {
...
 // use it within this method, as it does not run on the main thread and must
 // not run any method that would cause JavaScript to run. In practice, this
 // means that almost any use of napi_env will be incorrect.
-inline void AsyncWorker::OnExecute(napi_env /*DO_NOT_USE*/, void* this_pointer) {
-  AsyncWorker* self = static_cast<AsyncWorker*>(this_pointer);
+inline void AsyncWorker::OnExecute(Napi::Env /*DO_NOT_USE*/) {
 #ifdef NAPI_CPP_EXCEPTIONS
   try {
-    self->Execute();
+    this->Execute();
```
Comment (gabrielschulhof) [inline napi-inl.h]: Do we need `this->` here and in the lines below?

File: napi-inl.h [hunk 13]
```diff
@@ -4199,40 +4199,51 @@ inline void AsyncWorker::SetError(const std::string& error) {
...
+inline void AsyncWorker::OnAsyncWorkComplete(napi_env env,
+                                             napi_status status,
+                                             void* asyncworker) {
+  AsyncWorker* self = static_cast<AsyncWorker*>(asyncworker);
+  self->OnWorkComplete(env, status);
+}
+inline void AsyncWorker::OnWorkComplete(Napi::Env /*env*/, napi_status status) {
   if (status != napi_cancelled) {
-    HandleScope scope(self->_env);
+    HandleScope scope(this->_env);
```
Comment (gabrielschulhof) [inline napi-inl.h]: Same here. I don't think we need `this->`.

**Comment by legendecas** [comment]
@gabrielschulhof updated! PTAL :D

File: doc/async_worker.md [hunk 14]
```diff
@@ -136,6 +136,35 @@ class was created, passing in the error as the first parameter.
...
+virtual void OnWorkComplete(Napi::Env env, napi_status status);
+```
+
+### OnExecute
+
+This method is invoked immediately on the work thread when scheduled.
+The default implementation of this method just calls the `Napi::AsyncWorker::Execute`
+and handles exceptions if cpp exceptions were enabled.
+
+The `OnExecute` method receives an `napi_env` argument. However, do NOT
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
The `OnExecute` method receives an `napi_env` argument. However, it must NOT be used
```

File: doc/async_worker.md [hunk 15]
```diff
@@ -136,6 +136,35 @@ class was created, passing in the error as the first parameter.
...
+use it within this method, as it does not run on the JavaScript thread and
```
Comment (gabrielschulhof) [inline doc/async_worker.md]:
```suggestion
within this method, as it does not run on the JavaScript thread and
```

**Comment by gabrielschulhof** [comment]
CI:

| Version | Job | Status |
|---:|---|---|
| v14.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1541/ | ✓ |
| v12.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1542/ | ✓ |
| v10.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1543/ | ✓ |

**Comment by gabrielschulhof** [comment]
CI:

| Version | Job | Status |
|---:|---|:---:|
| v14.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1581/ | ✓ |
| v12.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1582/ | ✗ |
|  | https://ci.nodejs.org/job/node-test-node-addon-api-new/1584/ | ✗ |
|  | https://ci.nodejs.org/job/node-test-node-addon-api-new/1585/ | ✓ |
| v10.x | https://ci.nodejs.org/job/node-test-node-addon-api-new/1583/ | ✓ |

**Comment by gabrielschulhof** [comment]
Landed in 23ff7f0b245c23a01b30ffcf2fcf7b89c6faf2d9.

**Pull request closed without merging**